Optimize TransformationMatrix inverse on ARM64

cdumez opened the pull request:
#### cf6d576b811b032fc0322bb2b36020a1725c9a48
<pre>
Optimize TransformationMatrix inverse on ARM64
<a href="https://bugs.webkit.org/show_bug.cgi?id=252487">https://bugs.webkit.org/show_bug.cgi?id=252487</a>

Reviewed by Yusuke Suzuki.

Optimize TransformationMatrix inverse on ARM64 by using SIMD.

This is a cherry-pick of:
<a href="https://chromium.googlesource.com/chromium/blink/+/7b03a53c2c7f24fb1dd79c0e7f4b7fad520875a5">https://chromium.googlesource.com/chromium/blink/+/7b03a53c2c7f24fb1dd79c0e7f4b7fad520875a5</a>

I benchmarked this locally on M1 Pro with 100k matrices of random double values.
I saw a 20% speedup.

* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::adjoint):
(WebCore::inverse):

Canonical link: <a href="https://commits.webkit.org/260468@main">https://commits.webkit.org/260468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/19de229bbedf4c9a2dfa285012bcec2e76954e2a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108386 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41250 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117502 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116869 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112272 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8761 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100611 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114152 "Passed tests") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97416 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42145 "Passed tests") | 
| | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29061 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83835 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10313 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30405 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11056 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/7312 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16463 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50001 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7239 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12642 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
```diff
@@ -216,12 +213,129 @@ static bool inverse(const TransformationMatrix::Matrix4& matrix, TransformationM
         return false;
 
-    // Scale the adjoint matrix to get the inverse
+#if CPU(ARM64) && defined(_LP64)
```
Let's use `CPU(ARM64) && CPU(ADDRESS64)`

Constellation reviewed (approved):
r=me

cdumez committed:
Optimize TransformationMatrix inverse on ARM64
https://bugs.webkit.org/show_bug.cgi?id=252487

Reviewed by Yusuke Suzuki.

Optimize TransformationMatrix inverse on ARM64 by using SIMD.

This is a cherry-pick of:
https://chromium.googlesource.com/chromium/blink/+/7b03a53c2c7f24fb1dd79c0e7f4b7fad520875a5

I benchmarked this locally on M1 Pro with 100k matrices of random double values.
I saw a 20% speedup.

* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::adjoint):
(WebCore::inverse):

Canonical link: https://commits.webkit.org/260468@main
